matrix_sdk_base/src/models/message.rs

```diff
@@ -139,20 +139,38 @@ impl IntoIterator for MessageQueue {
 }
 
 pub(crate) mod ser_deser {
+    use std::fmt;
+
     use super::*;
 
+    struct MessageQueueDeserializer;
+
+    impl<'de> de::Visitor<'de> for MessageQueueDeserializer {
+        type Value = MessageQueue;
+
+        fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+            formatter.write_str("an array of message events")
+        }
+
+        fn visit_seq<S>(self, mut access: S) -> Result<Self::Value, S::Error>
+        where
+            S: de::SeqAccess<'de>,
+        {
+            let mut msgs = Vec::with_capacity(access.size_hint().unwrap_or(0));
```

Comment:
Would using `.unwrap_or(MESSAGE_QUEUE_CAP)` be even more optimized?

Reply:
So this is basically just taken from https://serde.rs/deserialize-map.html. I doubt `serde_json` will ever return something else than `Some(0)` or `None` (is this queue only used with JSON?). If we want to pre-allocate without knowing how many messages there are, we should probably just have `Vec::with_capacity(MESSAGE_QUEUE_CAP)`. This would ensure that the `Vec` never has to re-allocate but would waste space or needlessly allocate if the input contains fewer or no elements.

Depending on the allocator, `Vec::with_capacity(35 /* MESSAGE_QUEUE_CAP */)` might also not be a super good idea. I could *imagine* `32` being better. But this is where my expertise definitely ends.

I would estimate allocating up-front is not worth it unless there's basically always at least some messages, and there's more often 9+ messages than 0-8, but even then it might not be worth it. Some background:

* `Vec::with_capacity(0)` has capacity 0, doesn't allocate at all
* Pushing an element onto a 0-capacity `Vec` makes it allocate space for 4 elements (at least on Nightly, I think it used to go 0 -> 1 -> 2 -> 4 -> ... until quite recently)
* Pushing an element onto a full 4-capacity `Vec` makes it allocate space for 8 elements, move the existing 4 elements, and deallocate the smaller buffer
* Pushing an element onto a full 8-capacity `Vec` makes it allocate space for 16 elements, move the existing 8 element, and deallocate the smaller buffer